### ISSUE TITLE
Rotations seem to be non-deterministic and have unexpected values

### ISSUE THREAD

**Issue by RafaelMostert** [HITS-AIN/PINK#58]
(Using PINK Version 2.5 Git revision: feac562)
Using real data, mapping about 2000 sources to a 10x10 SOM I plotted the stored flips and rotations.
As expected, the flips are mostly random:
![flip_debug](https://user-images.githubusercontent.com/12486433/83876248-d9a99c00-a738-11ea-896c-c1604091f0fd.png)
The rotations are unexpectedly peaked at 4 different places, something that Tim also observed earlier.
![rotations_debug](https://user-images.githubusercontent.com/12486433/83876252-da423280-a738-11ea-84c7-644eb8d3c389.png)
Apart from this global effect, the reported rotations did not seem to have any relation to the individual inputted images.

Thus I tested a small SOM with mock data:
I created a 2x2 SOM with neurons that are all zeros except for the first neuron, which has a stripe pointing towards 12-oclock.
I created five images containing all zeros.
Except that the last four images have a stripe pointing towards 12, 3, 6 and 9 o-clock respectively.

I then map these five images to the 2x2 SOM and store the flips and rotations.
See attached ipynb:
[test_rotations_pink.zip](https://github.com/HITS-AIN/PINK/files/4736007/test_rotations_pink.zip)

Everytime I rerun the mapping command (without changing the input),
`Pink --cuda-off --store-rot-flip test_rotations.bin --euclidean-distance-type float --euclidean-distance-shape circular --som-width 2 --som-height 2 --map test_data.bin test_map.bin test_som.bin`
 I get reasonable and consistent values for the mapping distances from each image to each neuron.
However, the rotation angles and flip-bools are different everytime and they are never what I expect them to be.

The non-deterministic behaviour for the flips and rotations happen without cuda.
With cuda, I only get 0s for the rotation-angles in this test-case.

**Comment by tjgalvin** [comment]
I have not seen the non-deterministic behaviour you are describing before. I have seen the cluster of angles in increments of 90 degrees before. I checked out the git revision you linked and the bugs that caused that type of behaviour that I know of have already been patched. By chance, for the 2000 sources, are they often unresolved and point source like? I have seen some preference for such images to be close to rotations that are in increments of 90 degrees, which is mostly a result of the interpolation being most accurate at these increments. 

Out of curiosity, what happens when you run something similar to the following notebook? 

https://github.com/tjgalvin/pyink/blob/master/Example_Notebooks/Example_Angle_Comparison.ipynb

**Comment by RafaelMostert** [comment]
For 40 samples, for 50 data images.
(the same mock dataset as before only repeating the fifth object)

The things I changed in the script are the mask. I used a mask of 49 (as my image size is 100).
And I changed stuff to make it work for a single-channel SOM.

angles:
![image](https://user-images.githubusercontent.com/12486433/83901074-69ac0d80-a75a-11ea-80c6-a49e44f97539.png)

residual angles:
![image](https://user-images.githubusercontent.com/12486433/83901013-513bf300-a75a-11ea-9fbd-f434e4b674a1.png)


![image](https://user-images.githubusercontent.com/12486433/83893654-83485780-a750-11ea-9d3e-3deb41579631.png)

**Comment by tjgalvin** [comment]
Sorry, I am a little confused by what you mean by 'for 40 samples, for 50 data images'. May you clarify that for me? It looks to me like you are only ever getting five unique transformations returned, all of which have no flips. 

I had a quick look at your test notebook and I think there might be a problem with your function that creates the SOM binary. I loaded up the test SOM it generates with my own code and I got this header:

`(2, 1, 0, 0, 2, (2, 2), 0, 2, (2, 142))`

That `2` in the `(2, 142)` seems a bit off to me. That looks right to me

```
file.write(struct.pack('i' * 11, 2, #<file format version>
                                   1, #<file type>
                                   0, #<data type>
                                   0, #<som layout>
                                   2, 2,2, #<som dimensionality, width and height> 
                                   0, #<neuron layout>
                                   2, np.shape(som_data)[1], np.shape(som_data)[2])) #<neuron dimensionality and width> 
```

but printing the `som_data` shape in that function gives a shape of `(2, 2, 142, 142)`. I suspect that what you may want to do is ` np.shape(som_data)[2], np.shape(som_data)[3]` in the above. That would at least get the header information correct.

Now, I thought I remember reading some checks to see if the neuron dimensions where the same, but I can't seem to find them. @BerndDoser would know what the behavior is. What I thought I was remember was somewhere in `InputData.cpp`, but re-reading it now it looks like it forces square neurons by only accepting a single integer for it, and this check is only for user input. But, you may have neatly side-stepped this check by creating your own binary file.

I am not completely sure if this is right, but if PINK is not working with the same SOM you are in your test code I think is a little cloudy as to what is actually happening.  

Any thoughts @RafaelMostert ? I only had a quick look so I may be missing something. And this little discussion may be a little different from the original issue you had, which seemingly is about real data and a larger SOM.

**Comment by RafaelMostert** [comment]
Sorry for the bug there!
Indeed it needs to be `np.shape(som_data)[2], np.shape(som_data)[3]` I forgot to correct that in the uploaded zip. That did not alter the observed mapping behaviour for me. 

In any case, maybe you are right about it not addressing my initial concern: the rotation seemingly not matching my data.
I'll set up a simple test case with a real trained SOM produced by PINK and real data.

**Comment by tjgalvin** [comment]
I'm just a little unsure whether we are looking at one issue or two. How large are the binary files collectively? I might be able to help out with a closer look. 

I can say for my current data set that the spatial transformations are working. It did take a few attempts and a couple of bugs in the revised code base, but as far as I can tell it should be working.

What happens when you run the notebook I linked previous against the real data and not the simulated images you made? Also, in that pyink package there are some wrappers around the binary files that I am pretty confident with. It might be worth trying them out just to see if the behavior is the same.

**Comment by BerndDoser** [comment]
The file format of your SOM binary file in your jupyter notebook is not correct.

The script `show_som.py` shows:
```
Input file is: test_som.bin
Save location is: ./
The file name is: SOM
The border width is: 2
The background color is: #ffaadd
The intensity is shared: True
Display is off
version: 2
file_type: 1
data_type: 0
som_layout: 0
som dimensionality: 2
som dimensions: (2, 2)
neuron dimensionality: 2
neuron dimensions: (2, 142)
284 neurons loaded
True
cartesian map
```
Unfortunately, the neuron data layout has no extra layout index, as it was only Cartesian in the past. This is indeed very inconsistent and should be changed in a future release.

**Comment by BerndDoser** [comment]
Instead of changing the error-prone binary file format we should directly apply a better data format as discussed in #43.

**Comment by RafaelMostert** [comment]
@BerndDoser are you sure that the neuron data layout has no extra layout index?
Removing the neuron data layout in the SOM header gives weird results for show_som.py, while keeping it in works fine for show_som.py.

if I inspect the test SOM keeping it in (assuming  `np.shape(som_data)[2], np.shape(som_data)[3]` is fixed)
I get as output from your show_som.py:

```
Input file is: /data/mostertrij/pink-basics/output_LoTSS_DR1/test/test_som.bin
Save location is: ./
The file name is: SOM
The border width is: 2
The background color is: #ffaadd
The intensity is shared: True
Display is off
version: 2
file_type: 1
data_type: 0
som_layout: 0
som dimensionality: 2
som dimensions: (2, 2)
neuron dimensionality: 2
neuron dimensions: (142, 142)
4 neurons loaded
True
cartesian map
```
and the following image:
[SOM0.pdf](https://github.com/HITS-AIN/PINK/files/4745406/SOM0.pdf)
Which looks correct to me (except that it is flipped 90 degrees anti-clockwise).

**Comment by BerndDoser** [comment]
Yes,sorry @RafaelMostert. It seems to be simply a bug in the `show_som.py` script. I will go on to understand the issue.

**Comment by tjgalvin** [comment]
Off hand, is the rotated flip related to the column major / roe major
change that occurred from pink version 1 to version 2?

I haven't used that show some script before, but if it is meant to support
version 1 and version 2 file types there should also be an extra transpose
applied somewhere on a per neuron basis. Can you see it anywhere?

**Comment by RafaelMostert** [comment]
show_som.py does have a per neuron transform:

https://github.com/HITS-AIN/PINK/blob/feac56256d0b7272808326de9da4001b18f20904/scripts/show_som.py#L116-L117

**Comment by tjgalvin** [comment]
Ok - the new version of PINK is using C style notation now, so that extra transpose is not needed for version two PINK binary files. If it is hard-coded and there for all cases it is what is causing your extra flip in the earlier image.

**Comment by BerndDoser** [comment]
The files for SOM, map, and spatial transformations (rot_flip) are all written in the same order (as they are stored in memory).

**Comment by BerndDoser** [comment]
Now I understand more - I must be very tired. There is no bug in `show_som.py`. Fixing your notebook to `np.shape(som_data)[2], np.shape(som_data)[3]` looks better now. Your result is now:

```
Map result:
[[26.458  0.     0.     0.   ]
 [ 0.    26.178 26.178 26.178]
 [ 0.    26.178 26.178 26.178]
 [ 0.    26.178 26.178 26.178]
 [ 0.    26.178 26.178 26.178]] 5 2 2 1 

Neuron matching order per data input
[3 0 1 2]
[0 1 2 3]
[0 1 2 3]
[0 1 2 3]
[0 1 2 3]
Flips: [1 1 0 1 1]
Rotations [deg]: [0, 0, 270, 90, 180]
```

- Image 0 matched to neuron 1 (rotational invariant) - ok
- Image 1 matched neuron 0 (flipped, no rotation) - ok
- Image 2 matched neuron 0 (unflipped, 270 deg rotation) - ok
- Image 3 matched neuron 0 (flipped, 90 deg rotation) - not ok, should be 180 deg
- Image 4 matched neuron 0 (flipped, 180 deg rotation) - not ok, should be 270 deg

**Comment by BerndDoser** [comment]
The spatial transformation file is storing the information of each image to each neuron and not only to the best matching neuron. Therefore, you have to read an array of size `number_of_images * som_size` and take the best match index from mapping file.

**Comment by RafaelMostert** [comment]
Ah! That is good to know :)

From https://github.com/HITS-AIN/PINK/blob/master/FILE_FORMATS.md#best-rotation-and-flipping-parameter-file
I took that it was one flip and rotation per image:

> **Best rotation and flipping parameter file**
> 
> `<file format version> 3 <number of entries> <som layout> <data>`
> 
> The data section contains a bool (is flipped) and a 32-bit float number (angle in radian) for each neuron.

**Comment by tjgalvin** [comment]
@RafaelMostert  here is a link to the transformer class I have been using for my own stuff. If you'd like please feel free to use the package or borrow the class. 

https://github.com/tjgalvin/pyink/blob/d3c4f5329c76c6918c183c4cd759e601c6cb9871/pyink/binwrap.py#776

**Comment by BerndDoser** [comment]
However, there was indeed a race condition bug in the CPU path for the best rotation index.

Fixed with 6ebe6a681f9a228307e09e4beb975048db4ee486.

**Comment by BerndDoser** [comment]
I will close it for now. Please open a new issue if something is still unclear.
